crates/video/src/video.rs

```diff
@@ -0,0 +1,1089 @@
+                }
+            }
+            VIRTIO_VIDEO_CMD_QUEUE_CLEAR => {
+                let body: VideoGetParamsRequest = read_body!(true)?;
+                QueueClear {
+                    stream_id: body.hdr.stream_id.into(),
+                    queue_type: QueueType::n(<Le32 as Into<u32>>::into(body.queue_type)).ok_or(
+                        VuVideoError::UnexpectedArgValue(stringify!(body.queue_type).to_string()),
+                    )?,
+                }
+            }
+            VIRTIO_VIDEO_CMD_GET_PARAMS_EXT => {
+                let body: VideoGetParamsRequest = read_body!(true)?;
+                GetParams {
+                    stream_id: body.hdr.stream_id.into(),
+                    queue_type: QueueType::n(<Le32 as Into<u32>>::into(body.queue_type)).ok_or(
+                        VuVideoError::UnexpectedArgValue(stringify!(body.queue_type).to_string()),
+                    )?,
+                }
+            }
+            VIRTIO_VIDEO_CMD_SET_PARAMS_EXT => {
+                let body: VideoSetParamsRequest = read_body!(true)?;
+                SetParams {
+                    stream_id: body.hdr.stream_id.into(),
+                    queue_type: QueueType::n(<Le32 as Into<u32>>::into(body.params.queue_type))
+                        .ok_or(VuVideoError::UnexpectedArgValue(
+                            stringify!(body.queue_type).to_string(),
+                        ))?,
+                    params: body.params,
+                }
+            }
+            VIRTIO_VIDEO_CMD_GET_CONTROL => {
+                let body: VideoGetControlRequest = read_body!(true)?;
+                GetControl {
+                    stream_id: body.hdr.stream_id.into(),
+                    control: ControlType::n(<Le32 as Into<u32>>::into(body.control)).ok_or(
+                        VuVideoError::UnexpectedArgValue(stringify!(body.control).to_string()),
+                    )?,
+                }
+            }
+            _ => return Err(vhu_video::VuVideoError::InvalidCmdType(header.type_.into())),
+        })
+    }
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Default)]
+pub(crate) struct VirtioVideoHeader {
+    pub type_: Le32,
+    pub stream_id: Le32,
+}
+
+impl VirtioVideoHeader {
+    pub fn as_slice(&self) -> &[u8] {
+        unsafe {
+            return ::core::slice::from_raw_parts(
+                (self as *const VirtioVideoHeader) as *const u8,
+                ::core::mem::size_of::<VirtioVideoHeader>(),
+            );
+        };
+    }
+}
+
+unsafe impl ByteValued for VirtioVideoHeader {}
+
+/// Requests
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Default)]
+pub(crate) struct VideoQueryCapRequest {
+    pub hdr: VirtioVideoHeader,
+    pub queue_type: Le32,
+    padding: Le32,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, Default)]
+pub(crate) struct VideoQueryControlRequest {
+    pub hdr: VirtioVideoHeader,
+    pub control: Le32,
+    padding: Le32,
+    pub format: Le32,
+    padding_: Le32,
```

Comment:
Why is `_` a suffix? If it is unused, shouldn't it be a prefix? :thinking:

Reply:
Well, it is unused, but the `_` suffix was added to avoid name repetition.

In the driver it is defined as:

```c
struct virtio_video_query_control {
	struct virtio_video_cmd_hdr hdr;
	__le32 control; /* One of VIRTIO_VIDEO_CONTROL_* types */
	__u8 padding[4];
	/*
	 * Followed by a value of struct virtio_video_query_control_*
	 * in accordance with the value of control.
	 */
};

struct virtio_video_query_control_resp_profile {
	__le32 num;
	__u8 padding[4];
	/* Followed by an array le32 profiles[] */
};

[...]
```
I unified them in a single struct and had a duplicated padding field. For `type` (in previous structs), I used a suffix to avoid using a reserved name. So here I went for a suffix too to avoid duplication. But I am ok on making it a prefix, as it is also unused.

Reply:
Or I could also slipt them to follow kernel's definition.